extmod/modgzip.c: Add gzip.GzipFile providing compression and decompression.

jimmo opened the pull request:
Summary: Adds gzip/zlib/raw-deflate compression support. (See issues #7228, #5590, and earlier PRs #8195, #5613).

Historically MicroPython provided a CPython-compatible implementation of `zlib.decompress` as well as a MicroPython-specific `zlib.DecompIO` which is similar to CPython's `zlib.decompressobj` except more useful as a stream wrapper, for example to decompress a file or socket, because you pull (read) decompressed data out of it, rather than pushing (write) compressed data into it.

However, CPython also provides `gzip.GzipFile` which works exactly like MicroPython's `DecompIO`, the only limitation is that it only supports gzip-format data (whereas `DecompIO` supported the same `wbits` argument that `zlib.decompress` supported).

This PR removes the `zlib` module, and replaces it with a `gzip` module providing a CPython-compatible `gzip.GzipFile` (now supporting decompression _and_ compression), but also with a MicroPython-extension allowing `GzipFile` to read/write zlib and raw-deflate files (via an optional, MicroPython-specific `wbits` keyword argument).

For backwards compatibility, I have written a pure-Python implementation of the functionality of the former `zlib` module (including now `zlib.compress`), which will be published to micropython-lib (https://github.com/micropython/micropython-lib/pull/694) and can be optionally installed via mip (or frozen). It's only about 16 lines (351 byte .mpy file) but adds +428 bytes when frozen. The argument for not including it by default is:
- (De)compression isn't a feature that everyone uses.
- Even in applications that use (de)compression, GzipFile provides a much more memory-efficient way to work with reading/writing compressed data to/from a stream (which is the typical use case).

This PR includes the following changes:
- Add a shared `mp_stream___exit___obj` that avoids each file/stream-like object needing to implement this themselves. (This is a code size optimisation saving ~200 bytes)
- Add a memory-efficient implementation of LZ77 compression to lib/uzlib (Written by @dpgeorge ).
- General code size optimisations for lib/uzlib. (Saves around 650 bytes)
- Code cleanup and simplication for lib/uzlib.
- Remove the `zlib` module.
- Add the `gzip` module.
- Add tests for `gzip.GzipFile`.
- Update docs for `zlib` to describe the micropython-lib version which wraps `GzipFile`.
- Add docs for `gzip`.

Overall this PR adds +1076 bytes to PYBV11 (i.e. adding compression support). This is made up of -3012 (remove zlib), +2900 (add decompression-only GzipFile), +1188 (add compression to GzipFile).

Compared to #8195 & #5613 which are quite similar (cc @harbaum, @andrewleech) this PR adds support for streaming as well as non-gzip (i.e. zlib/raw) formats and a code size reduction as well as reduced memory usage in the compressor.

_This work was funded through GitHub Sponsors, by specific request from a supporter._

dpgeorge committed:
lib/uzlib: Add memory-efficient, streaming LZ77 compression support.

Signed-off-by: Damien George <damien@micropython.org>

jimmo committed:
lib/uzlib: Code size improvements.

- Replace 256-byte reverse-bits-in-byte lookup table with computation.
- Replace length and distance code lookup tables with computation.
- Remove comp_disabled check (it's unused).
- Make the dest_write_cb take the data pointer directly, rather than the
  Outbuf.

Saves 500 bytes on PYBV11.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modgzip.c: Add gzip module providing the GzipFile class.

This is a compatible implementation of CPython's gzip.GzipFile supporting
both compression and decompression of zlib and gzip files.

This class can also be used as a drop-in replacement for the zlib.DecompIO
class, as well as providing a building block (via io.BytesIO) to implement
zlib.decompress (and now, zlib.compress), supporting raw, zlib and gzip
files.

Enable/disable this on all ports/boards that zlib was previously configured
for.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/stream: Add mp_stream___exit___obj that calls mp_stream_close.

There are enough places that implement __exit__ by forwarding directly
to mp_stream_close that this saves code size.

For the cases where __exit__ is a no-op, additionally make their
MP_STREAM_CLOSE ioctl handled as a no-op.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Remove the zlib module.

This will be replaced with the gzip module providing the same
functionality, with an optional frozen Python wrapper providing a
replacement zlib module.

binascii.crc32 is temporarily disabled.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib: Clean up tinf->uzlib rename.

This library used a mix of "tinf" and "uzlib" to refer to itself.
Remove all use of "tinf" in the public API.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib: Combine zlib/gzip header parsing to allow auto-detect.

This supports `wbits` values between +40 to +47.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/gzip.rst: Add docs for gzip and gzip.GzipFile.

Also update zlib docs.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib/tinflate.c: Compact lookup tables.

Saves 68 bytes on PYBV11.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib/defl_static.c: Optimize zlib_start/finish_block.

- Collapsing the two adjacent calls to outbits saves 32 bytes.
- Bring defl_static.c into lz77.c allows better inlining, saves 24 bytes.

Merge the Outbuf/uzlib_lz77_state_t structs. Minor simplification that
doesn't change code size.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/extmod: Add tests for gzip.

These test GzipFile as well as zlib-compatible compress/decompress
wrappers.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
The main TODO is whether to enable compression by default. Currently it's set to the "extra features" level which is the same as decompression support. I would be quite happy to move it to "full features".

dpgeorge commented on docs/library/gzip.rst:
```diff
@@ -0,0 +1,84 @@
+Functions
+---------
+
+.. class:: GzipFile(fileobj, mode, wbits=25, hist=None, /)
```
I think this should be `GzipFile(*, fileobj, mode, wbits=25, hist=None)`

dpgeorge commented on extmod/modgzip.c:
```diff
@@ -0,0 +1,450 @@
+        { MP_QSTR_fileobj, MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_mode, MP_ARG_KW_ONLY | MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_wbits, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 25} }, // Default to gzip with smallest window.
+        { MP_QSTR_hist, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
The default should be `.u_rom_obj = MP_ROM_NONE` to match the docs signature.

robert-hh commented:
There is not much difference ATM between the EXTRA and FULL feature setting. It seems to be only detailed error reporting, and either setting produces the same code size for a SAMD51 and 432 byte difference for a Teensy 4.0. MIMXRT and SAMD51 are the only boards with FULL setting, whether needed or not.
Adding GZIP compression increases the code size for the boards by about 1k. That seems to be acceptable for boards which support EXTRA features, without having checked the code space exhausting for every possible port & board. Including GZIP in the FULL feature level set would at least avoid a possible code size conflict for the moment, but would require to adding at a per-port or per-board level is required.

dpgeorge commented:
It's worth noting that the compression implemented here uses much less memory compared to #8195 and #5613.  In particular it doesn't allocate a large hash table to store indices into the history of the previously seen text.  Instead it simply does a brute-force-search of the history text to find a match for the compressor.  Yes this is slower (linear search vs hash table lookup) but with a small enough history (eg 512 bytes) it's not that slow.  And a small history does not impact the compression too much.

To give some more concrete numbers comparing memory use between the approaches:
- standard approach (eg #8195): inplace compression, all text to compress must be in RAM (or at least memory addressable), and then an additional 16k bytes RAM of hash table pointers, pointing into the text
- the approach in this PR: streaming compression, only a limited amount of previous text must be in RAM (user selectable, defaults to 512 bytes)

So if you want to compress, say, 1k of data, the standard approach requires all that data to be in RAM, plus an additional 16k of RAM for the hash table pointers.  In this PR, you only need the 1k of data in RAM.  Or if it's streaming from a file (or elsewhere), you could get away with only 256 bytes of RAM for the sliding history and still get very decent compression.

In summary: because compression takes such a large amount of RAM (in the traditional case) and it's not really suitable for microcontrollers, the approach taken in this PR is to minimise RAM usage as much as possible, and still have acceptable performance (speed and compression ratio).

dpgeorge commented on extmod/modgzip.c:
```diff
@@ -0,0 +1,450 @@
+            // resulting output will include a zlib-specific header and trailer.
+            self->mode = GZIPFILE_MODE_WRITE_ZLIB;
+            self->state[0].write.input_checksum = 1; // ADLER32
+        } else if (*wbits >= 25 && *wbits <= 31) {
```
I reckon it's worth allowing a smaller minimum wbits, to get a smaller history.  Eg 22?  Then the 9's about could be 6's.

jimmo replied:
I think this is worthwhile as it's a significant memory saving for small inputs, even if it is another CPython diff.

I've implemented this in a separate commit, but it can be squashed later.

dpgeorge commented on lib/uzlib/lz77.c:
```diff
@@ -0,0 +1,109 @@
+// Compress the given chunk of data.
+void uzlib_lz77_compress(uzlib_lz77_state_t *state, const uint8_t *src, unsigned len) {
+    bool use_src_as_history = false;
+    if (state->hist_buf == NULL) {
```
This case is the in-place scenario, where the history is exactly the input text.

jimmo replied:
Updated. Removing this saves 64 bytes.

It was a neat feature for `zlib.compress` case, but I really think we should be encouraging use of the streaming interface everywhere (and I've tried to convey that in the docs).

jimmo committed:
lib/uzlib/lz77.c: Always use separate history buffer.

Because we only use the streaming source, this is just extra code size.

Saves 64 bytes on PYBV11.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modgzip.c: Allow for smaller window sizes than CPython.

Previously the wbits parameter matched CPython's allowed values exactly.

Now we support down to wbits=5 (i.e. 32 bytes) on both compress and
decompress. This allows for faster compression and much less memory
usage while still achieving some level of compression.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Here is a test to show that the changes to defl_static.c produce the exact same output (by calling `uzlib_literal` and `uzlib_match` with all possible inputs).

https://gist.github.com/jimmo/90b5bbc4ffc813890bc4e9231e22338f

jimmo commented:
Added more tests based on the coverage results.

jimmo commented:
Replaced by https://github.com/micropython/micropython/pull/11905

dpgeorge commented:
Closing in favour of #11905.
